### DISCUSSION TITLE
In a graph, should a reentry transition show only trigger specific entry methods?

### DISCUSSION THREAD

**Discussion by leeoades** [dotnet-state-machine/stateless#513] [Q&A]
Consider a state machine with states A and B.
- Trigger X transitions from A to B
- State B has 2 entry functions
  - OnEntry for any trigger called OnEveryEntry
  - OnEntryFrom(Trigger.X) called OnEntryFromX
- B allows Trigger X to reentry into B.

So from State A, on Trigger.X, we expect _OnEveryEntry_ and _OnEntryFromX_ to fire.
In State B, on Trigger X, we expect the same.

The Digraph generated is as such:

```
digraph {
compound=true;
node [shape=Mrecord]
rankdir="LR"
"A" [label="A"];
"B" [label="B|entry / OnEveryEntry"];

"A" -> "B" [style="solid", label="X / OnEntryFromX"];
"B" -> "B" [style="solid", label="X / OnEveryEntry"];
 init [label="", shape=point];
 init -> "A"[style = "solid"]
}

```

![graphviz (5)](https://user-images.githubusercontent.com/2321091/231213644-90c8debb-90ac-4b59-ae3f-84d625dbbb20.svg)

I would argue that the reentry transition currently labelled  

> _X / OnEveryEntry_  

would be more usefully labelled as 

> _X / OnEntryFromX_

The State B already displays _entry / OnEveryEntry_ to denote all entries.
The label on the X reentry repeats this information, and neglects to mention _OnEntryFromX_.

So I would expect this:
![graphviz (6)](https://user-images.githubusercontent.com/2321091/231216016-d758ddc4-88fe-40d9-b8aa-50723db5384b.svg)

Thoughts?

**Comment by leeoades** [comment]
A possibly cleaner representation would be to declare a conditional entry as such.

```
digraph {
compound=true;
node [shape=Mrecord]
rankdir="LR"
"A" [label="A"];
"B" [label="B|entry / OnEveryEntry\nentry[X] / OnEntryFromX"];

"A" -> "B" [style="solid", label="X"];
"B" -> "B" [style="solid", label="X"];
 init [label="", shape=point];
 init -> "A"[style = "solid"]
}

```

![graphviz (7)](https://user-images.githubusercontent.com/2321091/231216671-b6ae3f1d-6639-4f65-8204-ffc9639148a1.svg)

This cleans up the labels on the transition arrows quite nicely, whilst also grouping the entry functions together.

**Comment by mclift** [comment]
Hi @leeoades,

Well spotted, this is definitely not what we should expect. I can see some spurious logic in `GraphStyleBase.FormatAllTransitions`, where it selects the entry action: https://github.com/dotnet-state-machine/stateless/blob/dev/src/Stateless/Graph/GraphStyleBase.cs#L83. Could this also be the cause of issue #509?

Additionally, the re-entry action name gets dropped entirely if you remove the entry action for any trigger:

```
digraph {
compound=true;
node [shape=Mrecord]
rankdir="LR"
"A" [label="A"];
"B" [label="B"];

"A" -> "B" [style="solid", label="X / OnEntryFromX"];
"B" -> "B" [style="solid", label="X"];
 init [label="", shape=point];
 init -> "A"[style = "solid"]
}
```
![image](https://user-images.githubusercontent.com/2707465/232545092-76c68f89-6d9e-47d9-b6c7-7f40acf7a0ee.png)

Regarding the alternative representation you've proposed, this looks clearer to me too.

Do you have a fix in mind?

**Reply by leeoades** [reply]
@mclift - Wait? You want me to actually code it up too? 😂
Happy to give it a bash. Wanted to run the ideas past the collective first though as I'm a noob here.

**Reply by mclift** [reply]
> @mclift - Wait? You want me to actually code it up too? 😂 

😁 only if you'd like to!

I'll try to take a look myself soon otherwise.

**Reply by leeoades** [reply]
@mclift Regarding the fix to https://github.com/dotnet-state-machine/stateless/issues/509, I believe it is more correct to store the reference to the trigger so that it can be interrogated later rather than simply adding a check for it inside the delegate. 
https://github.com/dotnet-state-machine/stateless/pull/511/commits/f755ddfcc32072447752e4f0cb5848802bc7fd11#diff-0de579f20c3805bba5083d369cbe249e3df5328c0c35c8c029f7239c92db7d00 
Anyway, that's been merged in now so that's nice.
